BaseInterfaces/src/array.jl

```diff
@@ -0,0 +1,93 @@
+setindex!(A, v, i::Int)		                                                    (if IndexLinear) Scalar indexed assignment
+setindex!(A, v, I::Vararg{Int, N})                                              (if IndexCartesian, where N = ndims(A)) N-dimensional scalar indexed assignment
+getindex(A, I...)	                    defined in terms of scalar getindex	    Multidimensional and nonscalar indexing
+setindex!(A, X, I...)	                defined in terms of scalar setindex!	Multidimensional and nonscalar indexed assignment 
+iterate	                                defined in terms of scalar getindex	Iteration
+length(A)	                            prod(size(A))	                        Number of elements
+similar(A)	                            similar(A, eltype(A), size(A))	        Return a mutable array with the same shape and element type
+similar(A, ::Type{S})	                similar(A, S, size(A))	                Return a mutable array with the same shape and the specified element type
+similar(A, dims::Dims)	                similar(A, eltype(A), dims)	            Return a mutable array with the same element type and size dims
+similar(A, ::Type{S}, dims::Dims)	    Array{S}(undef, dims)	                Return a mutable array with the specified element type and size
+
+Non-traditional indices	                Default definition	                    Brief description
+axes(A)	                                map(OneTo, size(A))	                    Return a tuple of AbstractUnitRange{<:Integer} of valid indices
+similar(A, ::Type{S}, inds)	            similar(A, S, Base.to_shape(inds))	    Return a mutable array with the specified indices inds (see below)
+similar(T::Union{Type,Function}, inds)	T(Base.to_shape(inds))	                Return an array similar to T with the specified indices inds (see below)
+=#
+
+# And arbitrary new type for array values
+struct ArrayTestVal
+    a::Int
+end
+
+# In case `eltype` and `ndims` have custom methods
+# We should always be able to use these to mean the same thing
+_eltype(::AbstractArray{T}) where T = T
+_ndims(::AbstractArray{<:Any,N}) where N = N
+
+array_components = (;
+    mandatory = (;
+        type = A -> A isa AbstractArray,
+        eltype = (
+            A -> eltype(A) isa Type,
+            A -> eltype(A) == _eltype(A),
+        ),
+        ndims = (
+            A -> ndims(A) isa Int,
+            A -> ndims(A) == _ndims(A),
+        ),
+        size = (
+            A -> size(A) isa NTuple{<:Any,Int},
+            A -> length(size(A)) == ndims(A),
+        ),
+        getindex = (
+            A -> A[begin] isa eltype(A),
+            A -> A[map(first, axes(A))...] isa eltype(A),
+        ),
+        indexstyle = A -> IndexStyle(A) in (IndexCartesian(), IndexLinear()),
+    ),
+    # TODO implement all the optional conditions
+    optional = (;
+        setindex! = (
+            A -> length(A) > 1 || throw(ArgumentError("Test arrays must have more than one element to test setindex!")),
```

Comment:
I think a better property here is to check that setting any index from `eachindex` and retrieving that index again gives the same element. This isn't actually defined by Base though... In some contexts, this can be only temporarily true, e.g. if you have a `DistributedArray` that fails to propagate some information in an overlapping region due to a network failure, for example.